sentry-android-core/src/main/java/io/sentry/android/core/SentryStartupProfilingProvider.java

```diff
@@ -0,0 +1,147 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.Reader;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+@ApiStatus.Internal
+public final class SentryStartupProfilingProvider extends EmptySecureContentProvider {
+
+  private final @NotNull ILogger logger;
+  private final @NotNull BuildInfoProvider buildInfoProvider;
+
+  @TestOnly
+  SentryStartupProfilingProvider(
+      final @NotNull ILogger logger, final @NotNull BuildInfoProvider buildInfoProvider) {
+    this.logger = logger;
+    this.buildInfoProvider = buildInfoProvider;
+  }
+
+  public SentryStartupProfilingProvider() {
+    logger = new AndroidLogger();
+    buildInfoProvider = new BuildInfoProvider(logger);
+  }
+
+  @Override
+  public boolean onCreate() {
+    final @Nullable Context context = getContext();
+
+    if (context == null) {
+      logger.log(SentryLevel.FATAL, "App. Context from ContentProvider is null");
+      return false;
+    }
+
+    // Debug.startMethodTracingSampling() is only available since Lollipop
+    if (buildInfoProvider.getSdkInfoVersion() < Build.VERSION_CODES.LOLLIPOP) {
+      return false;
+    }
+
+    final @NotNull File cacheDir = AndroidOptionsInitializer.getCacheDir(context);
+    final @NotNull File configFile = new File(cacheDir, STARTUP_PROFILING_CONFIG_FILE_NAME);
+
+    // No config exists: startup profiling is not enabled
+    if (!configFile.exists() || !configFile.canRead()) {
+      return false;
+    }
+
+    try (final @NotNull Reader fileReader = new SentryFileReader(configFile)) {
+      final @Nullable SentryStartupProfilingOptions profilingOptions =
+          new JsonSerializer(SentryOptions.empty())
+              .deserialize(fileReader, SentryStartupProfilingOptions.class);
+
+      if (profilingOptions == null) {
+        logger.log(
+            SentryLevel.WARNING,
+            "Unable to deserialize the SentryStartupProfilingOptions. Startup profiling will not start.");
+        return false;
+      }
+
+      if (!profilingOptions.isProfilingEnabled()) {
+        logger.log(SentryLevel.INFO, "Profiling is not enabled. Startup profiling will not start.");
+        return false;
+      }
+
+      final @NotNull TracesSamplingDecision startupSamplingDecision =
+          new TracesSamplingDecision(
+              profilingOptions.isTraceSampled(),
+              profilingOptions.getTraceSampleRate(),
+              profilingOptions.isProfileSampled(),
+              profilingOptions.getProfileSampleRate());
+      // We store any sampling decision, so we can respect it when the first transaction starts
+      AppStartMetrics.getInstance().setStartupSamplingDecision(startupSamplingDecision);
+
+      if (!(startupSamplingDecision.getProfileSampled() && startupSamplingDecision.getSampled())) {
+        return false;
```

Comment:
nit: maybe add a log here as well?

Reply:
i don't know, we don't log when a transaction or profile is not sampled in general, but a debug log shouldn't be annoying 👍